{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "75b99003",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from matplotlib_venn import venn2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0b3b3d6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_significant(file_path):\n",
    "    significants = []\n",
    "    with open(file_path, \"r\") as f:\n",
    "        datas = f.readlines()\n",
    "        for data in datas[1:]:\n",
    "            data = data.split()\n",
    "            x, y = int(int(data[1])/1000), int(int(data[3])/1000)\n",
    "            p_value, q_value = float(data[5]), float(data[6])\n",
    "            concat = int(data[4])\n",
    "            significants.append((x, y, p_value, q_value, concat))\n",
    "    return significants\n",
    "\n",
    "def cal_significant(significants, proportion):\n",
    "    test = np.array(significants)\n",
    "    \n",
    "    def _filter(_values, max_nums, test):\n",
    "        values_sort_index = _values.argsort()\n",
    "        values_min_numbers_index = values_sort_index[max_nums-1]\n",
    "        p = max_nums\n",
    "        values_min_indexs = _values.argsort()[:p-1]\n",
    "        _test = test[values_min_indexs, :]\n",
    "        return _test\n",
    "\n",
    "    _p_values = test[:, 2]\n",
    "    _test = _filter(_p_values, int(_p_values.shape[0]*proportion), test)\n",
    "    _q_values = _test[:, 3]\n",
    "    _test = _filter(_q_values, int(_q_values.shape[0]*proportion), _test)\n",
    "    return _test\n",
    "\n",
    "def cal_significant_test(significants, proportion):\n",
    "    test = np.array(significants)\n",
    "\n",
    "    test = test[np.argsort(test[:, 2])]\n",
    "    _test = np.zeros((int(test.shape[0]*proportion), 5))\n",
    "    significant_set = set()\n",
    "    index = 0\n",
    "    for i in range(test.shape[0]):\n",
    "        x, y = int(test[i, 0]), int(test[i, 1])\n",
    "        flag = False\n",
    "        for _x in range(x-4, x+5):\n",
    "            for _y in range(y-4, y+5):\n",
    "                if (_x, _y) in significant_set:\n",
    "                    flag = True\n",
    "                    break\n",
    "            if flag:\n",
    "                break\n",
    "        if flag:\n",
    "            continue\n",
    "        significant_set.add((x, y))\n",
    "        _test[index] = test[i]\n",
    "        index += 1\n",
    "        if index > _test.shape[0] - 1:\n",
    "            break\n",
    "    \n",
    "    def _filter(_values, max_nums, test):\n",
    "        values_sort_index = _values.argsort()\n",
    "        values_min_numbers_index = values_sort_index[max_nums-1]\n",
    "        p = max_nums\n",
    "        values_min_indexs = _values.argsort()[:p-1]\n",
    "        _test = test[values_min_indexs, :]\n",
    "        return _test\n",
    "\n",
    "    _q_values = _test[:, 3]\n",
    "    _test = _filter(_q_values, int(_q_values.shape[0]*proportion), _test)\n",
    "    return _test\n",
    "\n",
    "def show_venn(enhanced_sig_set, hr_sig_set):\n",
    "    venn2(subsets = [hr_sig_set, enhanced_sig_set], set_labels = ('1kb', 'enhanced'), set_colors=((247/255, 129/255, 191/255), (51/255, 160/255, 44/255)))\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ce6bdc41",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/data1/lmh_data/MINE/GM12878_ATAC_H3K27ac_H3K4me3/analyse/GM12878_ATAC_H3K27ac_H3K4me3/experiment/loop/enhanced/chr1_1000b/outputs_2_12/FitHiC.spline_pass1.res1000.significances.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-369ed1db9205>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     14\u001b[0m     \u001b[0menhanced_file_path\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0menhanced_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'chr{}_1000b'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'outputs_2_12'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'FitHiC.spline_pass1.res1000.significances.txt'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m     \u001b[0mhr_file_path\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhr_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'chr{}_1000b'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'outputs'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'FitHiC.spline_pass1.res1000.significances.txt'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 16\u001b[0;31m     \u001b[0menhanced_sig\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcal_significant_test\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mget_significant\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0menhanced_file_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mproportion\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     17\u001b[0m     \u001b[0mhr_sig\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcal_significant\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mget_significant\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhr_file_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mproportion\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-2-799b0d1ccb60>\u001b[0m in \u001b[0;36mget_significant\u001b[0;34m(file_path)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mget_significant\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0msignificants\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m     \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"r\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m         \u001b[0mdatas\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadlines\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mdata\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdatas\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/data1/lmh_data/MINE/GM12878_ATAC_H3K27ac_H3K4me3/analyse/GM12878_ATAC_H3K27ac_H3K4me3/experiment/loop/enhanced/chr1_1000b/outputs_2_12/FitHiC.spline_pass1.res1000.significances.txt'"
     ]
    }
   ],
   "source": [
    "analyse_path = '/data1/lmh_data/MINE/GM12878_ATAC_H3K27ac_H3K4me3/analyse/GM12878_ATAC_H3K27ac_H3K4me3'\n",
    "enhanced_path = os.path.join(analyse_path, 'experiment', 'loop', 'enhanced')\n",
    "hr_path = os.path.join(analyse_path, 'experiment', 'loop', 'hr')\n",
    "auxiliary_files_path = os.path.join(analyse_path, 'experiment', 'auxiliary_files')\n",
    "proportion = 0.015\n",
    "\n",
    "# 未去交集的全部sig\n",
    "enhanced_sigs = set()\n",
    "hr_sigs = set()\n",
    "# 去交集的全部sig\n",
    "all_enhanced_sig = dict()\n",
    "all_hr_sig = dict()\n",
    "for chr in range(1, 23):\n",
    "    enhanced_file_path = os.path.join(enhanced_path, 'chr{}_1000b'.format(chr), 'outputs_2_100', 'FitHiC.spline_pass1.res1000.significances.txt')\n",
    "    hr_file_path = os.path.join(hr_path, 'chr{}_1000b'.format(chr), 'outputs_2_100', 'FitHiC.spline_pass1.res1000.significances.txt')\n",
    "    enhanced_sig = cal_significant_test(get_significant(enhanced_file_path), proportion)\n",
    "    hr_sig = cal_significant(get_significant(hr_file_path), proportion)\n",
    "    \n",
    "    _enhanced_sig = set(list(map(tuple, np.array(enhanced_sig)[:, :2])))\n",
    "    _hr_sig = set(list(map(tuple, np.array(hr_sig)[:, :2])))\n",
    "    show_venn(_enhanced_sig, _hr_sig)\n",
    "    # 去交集\n",
    "    new_enhanced_sig = _enhanced_sig - _hr_sig\n",
    "    new_hr_sig = _hr_sig - _enhanced_sig\n",
    "    \n",
    "    all_enhanced_sig[chr] = new_enhanced_sig\n",
    "    all_hr_sig[chr] = new_hr_sig\n",
    "    \n",
    "    for sig in _enhanced_sig:\n",
    "        _sig = list(sig)\n",
    "        _sig.append('chr{}'.format(chr))\n",
    "        enhanced_sigs.add(tuple(_sig))\n",
    "    for sig in _hr_sig:\n",
    "        _sig = list(sig)\n",
    "        _sig.append('chr{}'.format(chr))\n",
    "        hr_sigs.add(tuple(_sig))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fa79f05",
   "metadata": {},
   "outputs": [],
   "source": [
    "show_venn(enhanced_sigs, hr_sigs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea911919",
   "metadata": {},
   "outputs": [],
   "source": [
    "attentions = 50\n",
    "center = attentions - 1\n",
    "\n",
    "def read_chiqseq_file(path):\n",
    "    _peaks = dict()\n",
    "    with open(path)as f:\n",
    "        for line in f:\n",
    "            datas = line.strip().split()\n",
    "            left, right = int(int(datas[1])/1000), int(int(datas[2])/1000)\n",
    "            peaks = list(range(left, right+1))\n",
    "            if datas[0] not in _peaks.keys():\n",
    "                _peaks[datas[0]] = set()\n",
    "            for i in peaks:\n",
    "                _peaks[datas[0]].add(i)\n",
    "    return _peaks\n",
    "\n",
    "def get_hit_indicators(significants, peaks):\n",
    "    distance = np.zeros(attentions*2-1)\n",
    "    \n",
    "    for sig in significants:\n",
    "        x, y = int(sig[0]), int(sig[1])\n",
    "        for i in range(attentions):\n",
    "            if (x+i) in peaks or (y+i) in peaks:\n",
    "                distance[center-i] += 1\n",
    "            if i != 0 and((x-i) in peaks or (y-i) in peaks):\n",
    "                distance[center+i] += 1\n",
    "            if (x+i) in peaks or (x-i) in peaks or (y+i) in peaks or (y-i) in peaks:\n",
    "                break\n",
    "    return distance\n",
    "\n",
    "def show(x, mine_result, hr_result):\n",
    "    sns.set_style(\"whitegrid\") \n",
    "    fig, axes = plt.subplots(nrows=1, ncols=1, figsize=(5, 5), sharey=True)\n",
    "    axes.grid(False)\n",
    "    axes.plot(x, mine_result, color=(51/255, 160/255, 44/255), label='mine')\n",
    "    axes.plot(x, hr_result, color=(177/255, 89/255, 40/255), label='hr')\n",
    "\n",
    "    plt.xlabel('distance')\n",
    "    plt.ylabel('nums')\n",
    "    plt.legend()\n",
    "    plt.show()\n",
    "\n",
    "def figure(chip_seq_path):\n",
    "    all_enhanced_result = np.zeros(attentions*2-1)\n",
    "    all_hr_result = np.zeros(attentions*2-1)\n",
    "\n",
    "    for chr in range(1, 23):\n",
    "        _peaks = read_chiqseq_file(chip_seq_path)\n",
    "        enhanced_result = get_hit_indicators(all_enhanced_sig[chr], _peaks['chr{}'.format(chr)])\n",
    "        hr_result = get_hit_indicators(all_hr_sig[chr], _peaks['chr{}'.format(chr)])\n",
    "\n",
    "        all_enhanced_result += enhanced_result\n",
    "        all_hr_result += hr_result\n",
    "\n",
    "    show(list(range(-attentions+1, attentions)), all_enhanced_result, all_hr_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af9f960b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d50b7c24",
   "metadata": {},
   "outputs": [],
   "source": [
    "tss_file_path = os.path.join(auxiliary_files_path, 'Homo_sapiens.GRCh38.104.chr.gff3')\n",
    "\n",
    "def get_tss_gene_info(gene_biotype='protein_coding'):\n",
    "    gene_location = dict()\n",
    "    gene_info = dict()\n",
    "    with open(tss_file_path, \"r\") as f:\n",
    "        datas = f.readlines()\n",
    "        for data in datas[9:]:\n",
    "            data = data.split('\\t')\n",
    "            if len(data) < 9:\n",
    "                continue\n",
    "            if data[2] != 'gene' or data[8].find(gene_biotype) == -1:\n",
    "                continue\n",
    "            if data[6] == '+':\n",
    "                _location = data[3]\n",
    "            elif data[6] == '-':\n",
    "                _location = data[4]\n",
    "            else:\n",
    "                print('error')\n",
    "            if not data[0].isdigit():\n",
    "                continue\n",
    "            chromosome = int(data[0])\n",
    "            if chromosome not in gene_location:\n",
    "                gene_location[chromosome] = set()\n",
    "            gene_location[chromosome].add(int(int(_location)/1000))\n",
    "            \n",
    "            _tmp = data[-1].split('Name=')\n",
    "            if len(_tmp) < 2:\n",
    "                continue\n",
    "            gene_names = _tmp[1].split(';')[0]\n",
    "            gene_ids = data[-1].split('ID=')[1].split(',')[0].split(';')[0]\n",
    "            \n",
    "            if chromosome not in gene_info:\n",
    "                gene_info[chromosome] = set()\n",
    "            gene_info[chromosome].add((int(int(_location)/1000), gene_names, gene_ids, data[-1].replace('\\n', '')))\n",
    "    return gene_location, gene_info\n",
    "\n",
    "def figure_tss(locations):\n",
    "    all_enhanced_result = np.zeros(attentions*2-1)\n",
    "    all_hr_result = np.zeros(attentions*2-1)\n",
    "\n",
    "    for chr in range(1, 23):\n",
    "        if chr not in locations:\n",
    "            continue\n",
    "        enhanced_result = get_hit_indicators(all_enhanced_sig[chr], locations[chr])\n",
    "        hr_result = get_hit_indicators(all_hr_sig[chr], locations[chr])\n",
    "\n",
    "        all_enhanced_result += enhanced_result\n",
    "        all_hr_result += hr_result\n",
    "\n",
    "    show(list(range(-attentions+1, attentions)), all_enhanced_result, all_hr_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdd66dd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "figure_tss(get_tss_gene_info('protein_coding')[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78d79658",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "figure_tss(get_tss_gene_info('pseudogene')[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c95013e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_hit_genes(significants, gene_info, genes_loop_nums):\n",
    "    genes = set()\n",
    "    for sig in significants:\n",
    "        x, y = int(sig[0]), int(sig[1])\n",
    "        for _info in gene_info:\n",
    "            if abs(x-_info[0])<3 or abs(y-_info[0])<3:\n",
    "                if _info[1] not in genes_loop_nums:\n",
    "                    genes_loop_nums[_info[1]] = 1\n",
    "                else:\n",
    "                    genes_loop_nums[_info[1]] += 1\n",
    "                genes.add((_info[1], _info[2], _info[3]))\n",
    "    return genes, genes_loop_nums\n",
    "\n",
    "def get_hit_gene_names(significants, gene_biotype='protein_coding'):\n",
    "    gene_info = get_tss_gene_info(gene_biotype)[1]\n",
    "    all_genes = set()\n",
    "    genes_loop_nums = dict()\n",
    "    for chr in range(1, 23):\n",
    "        genes, genes_loop_nums = get_hit_genes(significants[chr], gene_info[chr], genes_loop_nums)\n",
    "        all_genes = all_genes | genes\n",
    "    return all_genes, genes_loop_nums"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f437afd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "mine_hit_genes, mine_genes_loop_nums = get_hit_gene_names(all_enhanced_sig, gene_biotype='protein_coding')\n",
    "mine_hit_genes = list(mine_hit_genes)\n",
    "mine_hit_genes = np.array(mine_hit_genes)\n",
    "print(len(mine_hit_genes))\n",
    "print(mine_genes_loop_nums)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c9c8e13",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = list(mine_genes_loop_nums.keys())\n",
    "y = list(mine_genes_loop_nums.values())\n",
    "\n",
    "sns.set_style(\"whitegrid\") \n",
    "fig, axes = plt.subplots(nrows=1, ncols=1, sharey=True)\n",
    "axes.grid(False)\n",
    "axes.plot(x, y, color=(166/255, 206/255, 227/255), label='mine')\n",
    "\n",
    "plt.xticks(rotation=90, fontsize=14)\n",
    "\n",
    "# plt.ylim(0, 1)\n",
    "plt.xlabel('gene_names')\n",
    "plt.ylabel('nums')\n",
    "# plt.xlim(0, len(significants))\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afb84b27",
   "metadata": {},
   "outputs": [],
   "source": [
    "hr_hit_genes, hr_genes_loop_nums = get_hit_gene_names(all_hr_sig, gene_biotype='protein_coding')\n",
    "hr_hit_genes = list(hr_hit_genes)\n",
    "hr_hit_genes = np.array(hr_hit_genes)\n",
    "print(len(hr_hit_genes))\n",
    "print(hr_genes_loop_nums)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7d92c75",
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import ticker\n",
    "x = list(hr_genes_loop_nums.keys())\n",
    "y = list(hr_genes_loop_nums.values())\n",
    "\n",
    "sns.set_style(\"whitegrid\") \n",
    "fig, axes = plt.subplots(nrows=1, ncols=1, sharey=True)\n",
    "axes.grid(False)\n",
    "axes.plot(x, y, color=(166/255, 206/255, 227/255), label='hr')\n",
    "\n",
    "# axes.xaxis.set_major_locator(ticker.MultipleLocator(4))\n",
    "plt.xticks(rotation=90, fontsize=10)\n",
    "\n",
    "# plt.ylim(0, 1)\n",
    "plt.xlabel('gene_names', fontsize=10)\n",
    "plt.ylabel('nums')\n",
    "# plt.xlim(0, len(significants))\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "MMSR",
   "language": "python",
   "name": "mmsr"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
