{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e046d82c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "from scipy.spatial import ConvexHull"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4937dc1c",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'np' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-8269c3815830>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# enhanced loop\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m HepG2_ATAC_H3K27ac_H3K4me3_2_100_all_enhanced_sig = np.load(\n\u001b[0m\u001b[1;32m      3\u001b[0m     \u001b[0;34m'/home/lmh/work/repos/MINE/analyse/fig3/represses/tss/temp/HepG2_H3K9me3_H3K27me3_2_100_all_enhanced_sig.npy'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0mallow_pickle\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m ).item()\n",
      "\u001b[0;31mNameError\u001b[0m: name 'np' is not defined"
     ]
    }
   ],
   "source": [
    "# enhanced loop\n",
    "HepG2_ATAC_H3K27ac_H3K4me3_2_100_all_enhanced_sig = np.load(\n",
    "    '/home/lmh/work/repos/MINE/analyse/fig3/represses/tss/temp/HepG2_H3K9me3_H3K27me3_2_100_all_enhanced_sig.npy',\n",
    "    allow_pickle=True\n",
    ").item()\n",
    "\n",
    "# rpkms info\n",
    "Homo_sapiens_GRCh38_file = '/data1/lmh_data/MINE/source/Homo_sapiens.GRCh38.84.gtf'\n",
    "Homo_sapiens_GRCh38 = dict()\n",
    "with open(Homo_sapiens_GRCh38_file, 'r')as f:\n",
    "    for line in f:\n",
    "        datas = line.strip().split()\n",
    "        if len(datas) < 4:\n",
    "            continue\n",
    "        if datas[2] != \"gene\":\n",
    "            continue\n",
    "        gene_id = datas[9].split('\"')[1]\n",
    "        Homo_sapiens_GRCh38[gene_id] = [datas[0], int(datas[3]), int(datas[4])]\n",
    "RNA_seq_file = '/data1/lmh_data/MINE/source/HepG2/RNA_seq_rpkms.xls'\n",
    "rna_infos = pd.read_excel(RNA_seq_file)\n",
    "rna_infos[\"rna_rpkms_avg\"] = rna_infos[[\"SS_FSU_43_none_RNAseq\", \"SS_FSU_44_none_RNAseq\"]].mean(axis=1)\n",
    "rna_infos = rna_infos.loc[rna_infos[\"rna_rpkms_avg\"]>0]\n",
    "rna_infos.insert(rna_infos.shape[1], 'Chromosome', '')\n",
    "rna_infos.insert(rna_infos.shape[1], 'Gene start (bp)', 0)\n",
    "rna_infos.insert(rna_infos.shape[1], 'Gene end (bp)', 0)\n",
    "for index in rna_infos.index:\n",
    "    if rna_infos.loc[index][\"Geneid\"] not in Homo_sapiens_GRCh38.keys():\n",
    "        continue\n",
    "    data = Homo_sapiens_GRCh38[rna_infos.loc[index][\"Geneid\"]]\n",
    "    rna_infos.loc[index, \"Chromosome\"] = data[0]\n",
    "    rna_infos.loc[index, \"Gene start (bp)\"] = data[1]\n",
    "    rna_infos.loc[index, \"Gene end (bp)\"] = data[2]\n",
    "\n",
    "# seq file path\n",
    "auxiliary_files_path = '/data1/lmh_data/MINE/source/HepG2'\n",
    "CTCF_path = os.path.join(auxiliary_files_path, 'CTCF_peaks.bed')\n",
    "POLR2A_path = os.path.join(auxiliary_files_path, 'POLR2A_peaks.bed')\n",
    "EP300_path = os.path.join(auxiliary_files_path, 'EP300_peaks.bed')\n",
    "EZH2_path = os.path.join(auxiliary_files_path, 'EZH2_peaks.bed')\n",
    "\n",
    "H3K27me3_path = os.path.join(auxiliary_files_path, 'H3K27me3_peaks.bed')\n",
    "H3K9me3_path = os.path.join(auxiliary_files_path, 'H3K9me3_peaks.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0fde403",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_score(sigs, _chr, _start, _end):\n",
    "    length = int((_end - _start) / 10 + 1)\n",
    "    scores = np.zeros(length)\n",
    "    for loop in sigs[_chr]:\n",
    "        for anchor in loop:\n",
    "            if anchor >= _start and anchor < _end + 10:\n",
    "                scores[int((anchor-_start)/10)] += 1\n",
    "    return scores\n",
    "\n",
    "def get_rpkms_score(infos, _chr, _start, _end):\n",
    "    length = int((_end - _start) / 10 + 1)\n",
    "    scores = np.zeros(length)\n",
    "    temp = infos.loc[infos[\"Chromosome\"]==str(_chr)]\n",
    "    for i in range(length):\n",
    "        tmp = temp.loc[temp[\"Gene start (bp)\"]>=(_start+i*10)*1000].loc[temp[\"Gene end (bp)\"]<(_start+(i+1)*10)*1000]\n",
    "        scores[i] = tmp[[\"rna_rpkms_avg\"]].sum()\n",
    "    return scores\n",
    "\n",
    "def get_seq_score(chip_seq_path, _chr, _start, _end):\n",
    "    length = int((_end - _start) / 10 + 1)\n",
    "    scores = np.zeros(length)\n",
    "    with open(chip_seq_path)as f:\n",
    "        for line in f:\n",
    "            datas = line.strip().split()\n",
    "            left, right = int(int(datas[1])/1000), int(int(datas[2])/1000)\n",
    "            if left >= _start and right < _end + 10:\n",
    "                scores[int((left-_start)/10)] += 1\n",
    "    return scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2563606",
   "metadata": {},
   "outputs": [],
   "source": [
    "_ranges = {\n",
    "    'represses_1961': [5, 39570, 52660],\n",
    "    'represses_2084': [5, 175460, 179160],\n",
    "    'represses_2497': [9, 6340, 12710],\n",
    "    'represses_1178': [18, 36820, 45660] \n",
    "}\n",
    "\n",
    "PASTIS_out_path = '/data1/lmh_data/MINE/GM12878_H3K9me3_H3K27me3/analyse/HepG2_H3K9me3_H3K27me3/experiment/SDOC/out_with_represses_loop_anchor/HepG2/PASTIS_out'\n",
    "def read_PASTIS_out(file_path):\n",
    "    _p = []\n",
    "    with open(file_path, \"r\") as f:\n",
    "        datas = f.readlines()\n",
    "        for data in datas:\n",
    "            _datas = data.split()\n",
    "            _data = [np.nan if d == 'nan' else float(d) for d in _datas]\n",
    "            _p.append(_data)\n",
    "    return _p\n",
    "\n",
    "points = dict()\n",
    "for key in _ranges.keys():\n",
    "    _range = _ranges[key]\n",
    "    file_name = 'PM2.chr{}_{}000_{}000'.format(_range[0], _range[1], _range[2])\n",
    "\n",
    "    points[key] = np.array(read_PASTIS_out(os.path.join(PASTIS_out_path, file_name)))\n",
    "    \n",
    "represses_1961 = points['represses_1961']\n",
    "represses_2084 = points['represses_2084']\n",
    "represses_2497 = points['represses_2497']\n",
    "represses_1178 = points['represses_1178']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2439956",
   "metadata": {},
   "outputs": [],
   "source": [
    "def show(points, scores, title):\n",
    "    very_large = 24; large = 22; med = 16; small = 12\n",
    "    params = {'axes.titlesize': very_large,\n",
    "              'legend.fontsize': med,\n",
    "              'figure.figsize': (10, 10),\n",
    "              'axes.labelsize': large,\n",
    "              'axes.titlesize': large,\n",
    "              'xtick.labelsize': large,\n",
    "              'ytick.labelsize': large,\n",
    "              'figure.titlesize': very_large,\n",
    "              'lines.linewidth': 2}\n",
    "    plt.rcParams.update(params)\n",
    "    \n",
    "    font = {'family': 'sans-serif', 'sans-serif': 'Arial'}\n",
    "    plt.rc('font', **font)\n",
    "\n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_subplot(111, projection=\"3d\")\n",
    "    cm = plt.get_cmap(\"Reds\")\n",
    "    col = [cm(float(i)/max(scores)) for i in scores]\n",
    "\n",
    "    ax.plot(points[:, 0], points[:, 1], points[:, 2], '-')\n",
    "    ax.scatter(points[:, 0], points[:, 1], points[:, 2], c=col)\n",
    "    ax.set_title(title)\n",
    "\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28021014",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "_range, _points = _ranges['represses_1961'], represses_1961\n",
    "scores = get_score(HepG2_ATAC_H3K27ac_H3K4me3_2_100_all_enhanced_sig, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'anchor(represses ultra high)')\n",
    "scores = get_rpkms_score(rna_infos, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'rpkms(represses ultra high)')\n",
    "\n",
    "scores = get_seq_score(CTCF_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'CTCF(represses ultra high)')\n",
    "scores = get_seq_score(POLR2A_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'POLR2A(represses ultra high)')\n",
    "scores = get_seq_score(EP300_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'EP300(represses ultra high)')\n",
    "scores = get_seq_score(EZH2_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'EZH2(represses ultra high)')\n",
    "\n",
    "scores = get_seq_score(H3K27me3_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'H3K27me3(represses ultra high)')\n",
    "scores = get_seq_score(H3K9me3_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'H3K9me3(represses ultra high)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c599b2c7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "_range, _points = _ranges['represses_2084'], represses_2084\n",
    "scores = get_score(HepG2_ATAC_H3K27ac_H3K4me3_2_100_all_enhanced_sig, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'anchor(represses high)')\n",
    "scores = get_rpkms_score(rna_infos, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'rpkms(represses high)')\n",
    "\n",
    "scores = get_seq_score(CTCF_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'CTCF(represses high)')\n",
    "scores = get_seq_score(POLR2A_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'POLR2A(represses high)')\n",
    "scores = get_seq_score(EP300_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'EP300(represses high)')\n",
    "scores = get_seq_score(EZH2_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'EZH2(represses high)')\n",
    "\n",
    "scores = get_seq_score(H3K27me3_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'H3K27me3(represses high)')\n",
    "scores = get_seq_score(H3K9me3_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'H3K9me3(represses high)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbcdefa9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "_range, _points = _ranges['represses_2497'], represses_2497\n",
    "scores = get_score(HepG2_ATAC_H3K27ac_H3K4me3_2_100_all_enhanced_sig, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'anchor(represses middle)')\n",
    "scores = get_rpkms_score(rna_infos, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'rpkms(represses middle)')\n",
    "\n",
    "scores = get_seq_score(CTCF_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'CTCF(represses middle)')\n",
    "scores = get_seq_score(POLR2A_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'POLR2A(represses middle)')\n",
    "scores = get_seq_score(EP300_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'EP300(represses middle)')\n",
    "scores = get_seq_score(EZH2_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'EZH2(represses middle)')\n",
    "\n",
    "scores = get_seq_score(H3K27me3_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'H3K27me3(represses middle)')\n",
    "scores = get_seq_score(H3K9me3_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'H3K9me3(represses middle)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccf685d9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "_range, _points = _ranges['represses_1178'], represses_1178\n",
    "scores = get_score(HepG2_ATAC_H3K27ac_H3K4me3_2_100_all_enhanced_sig, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'anchor(represses low)')\n",
    "scores = get_rpkms_score(rna_infos, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'rpkms(represses low)')\n",
    "\n",
    "scores = get_seq_score(CTCF_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'CTCF(represses low)')\n",
    "scores = get_seq_score(POLR2A_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'POLR2A(represses low)')\n",
    "scores = get_seq_score(EP300_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'EP300(represses low)')\n",
    "scores = get_seq_score(EZH2_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'EZH2(represses low)')\n",
    "\n",
    "scores = get_seq_score(H3K27me3_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'H3K27me3(represses low)')\n",
    "scores = get_seq_score(H3K9me3_path, _range[0], _range[1], _range[2])\n",
    "show(_points, scores, 'H3K9me3(represses low)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c96a5413",
   "metadata": {},
   "outputs": [],
   "source": [
    "def show(points, title):\n",
    "    pts = np.array(points)\n",
    "    mask = np.any(np.isnan(pts), axis=1)\n",
    "    pts = pts[~mask]\n",
    "\n",
    "    hull = ConvexHull(pts)\n",
    "\n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_subplot(111, projection=\"3d\")\n",
    "\n",
    "    # 12 = 2 * 6 faces are the simplices (2 simplices per square face)\n",
    "    for s in hull.simplices:\n",
    "        s = np.append(s, s[0])  # Here we cycle back to the first coordinate\n",
    "        ax.plot(pts[s, 0], pts[s, 1], pts[s, 2], \"r-\")\n",
    "    ax.set_title(title)\n",
    "\n",
    "    # Make axis label\n",
    "    for i in [\"x\", \"y\", \"z\"]:\n",
    "        eval(\"ax.set_{:s}label('{:s}')\".format(i, i))\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "show(represses_1961, 'ConvexHull(represses ultra high)')\n",
    "show(represses_2084, 'ConvexHull(represses high)')\n",
    "show(represses_2497, 'ConvexHull(represses middle)')\n",
    "show(represses_1178, 'ConvexHull(represses low)')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "MMSR",
   "language": "python",
   "name": "mmsr"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
